File: Strivers-A2Z-DSA-Sheet/04.Linked List/2.Doubly Linked List/javascript/01.Introduction_to_Double_LL.js
/*
QUESTION:
Construct a doubly linked list from an array and return the head.
Each node has three fields: data, prev (pointer to previous), next (pointer to next).

Example 1:
  Input:  arr = [1, 2, 3, 4, 5]
  Output: 1 <--> 2 <--> 3 <--> 4 <--> 5

Example 2:
  Input:  arr = [3, 4, 5]
  Output: 3 <--> 4 <--> 5

APPROACH:
1. Create head node from arr[0], set head.prev = null.
2. Use a curr pointer starting at head.
3. For each remaining element, create a new node:
   - new.prev = curr
   - curr.next = new
4. Advance curr to the new node.
5. Return head.

Time Complexity:  O(N)
Space Complexity: O(N)
*/

class Node {
    constructor(data) {
        this.data = data;
        this.prev = null;
        this.next = null;
    }
}

function constructDLL(arr) {
    if (arr.length === 0) return null;
    const head = new Node(arr[0]);
    let curr = head;
    for (let i = 1; i < arr.length; i++) {
        const newNode = new Node(arr[i]);
        newNode.prev = curr;   // link back
        curr.next = newNode;   // link forward
        curr = newNode;
    }
    return head;
}

// Helper: print DLL forward
function printDLL(head) {
    const result = [];
    let curr = head;
    while (curr) {
        result.push(curr.data);
        curr = curr.next;
    }
    console.log(result.join(' <--> '));
}

// Driver code
const head = constructDLL([1, 2, 3, 4, 5]);
printDLL(head); // 1 <--> 2 <--> 3 <--> 4 <--> 5

File: Strivers-A2Z-DSA-Sheet/04.Linked List/2.Doubly Linked List/python/01.Introduction_to_Double_LL.py
"""
QUESTION:
Construct a doubly linked list from an array and return the head.
Each node has three fields: data, prev (pointer to previous), next (pointer to next).

Example 1:
  Input:  arr = [1, 2, 3, 4, 5]
  Output: 1 <--> 2 <--> 3 <--> 4 <--> 5

Example 2:
  Input:  arr = [3, 4, 5]
  Output: 3 <--> 4 <--> 5

APPROACH:
1. Create head node from arr[0], set head.prev = None.
2. Use a curr pointer starting at head.
3. For each remaining element, create a new node:
   - new_node.prev = curr
   - curr.next = new_node
4. Advance curr to the new node.
5. Return head.

Time Complexity:  O(N)
Space Complexity: O(N)
"""


class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None


def construct_dll(arr):
    if not arr:
        return None
    head = Node(arr[0])
    curr = head
    for i in range(1, len(arr)):
        new_node = Node(arr[i])
        new_node.prev = curr    # link back
        curr.next = new_node    # link forward
        curr = new_node
    return head


# Helper: print DLL forward
def print_dll(head):
    result = []
    curr = head
    while curr:
        result.append(str(curr.data))
        curr = curr.next
    print(" <--> ".join(result))


# Driver code
if __name__ == "__main__":
    head = construct_dll([1, 2, 3, 4, 5])
    print_dll(head)  # 1 <--> 2 <--> 3 <--> 4 <--> 5

    head2 = construct_dll([3, 4, 5])
    print_dll(head2)  # 3 <--> 4 <--> 5

